### ISSUE TITLE
Continue on Multithreaded break shows other threads (besides main thread) still paused in UI

### ISSUE THREAD

**Issue by pieandcakes** [Microsoft/vscode#18195]
- VSCode Version: 1.8.1
- OS Version: Ubuntu 14.04

Steps to Reproduce:

1. Install ms-cpptools extension 0.9.3
2. Create a multithreaded app
3. Set a breakpoint where all threads have been started
4. hit breakpoint and confirm in Call Stack window that it shows "Pause on Breakpoint"
5. hit continue -> all threads should continue, (which is what happens) but the UI still shows all threads besides main thread as "Pause on Breakpoint"

I have confirmed that we send an empty body on `ContinueResponse` but I have also tried sending `"allThreadsContinued": true` in the body with no change. The comment says if the value is not set it defaults to true.

**Comment by sandy081** [comment]
@weinand  Not sure if this is related to our debugging infrastructure or an issue in the cpp debugging adapter hence assigning to you to decide.

**Comment by isidorn** [comment]
@pieandcakes it would be cool if your team was using vscode insiders so we would get feedback faster in case we break something.

**Comment by isidorn** [comment]
ping @weinand to consider introducing allThreadsContinued as a capability

**Comment by pieandcakes** [comment]
@isidorn I will try and do that. I'm currently the only one working on our extension so a lot of bugs are reaction to customer issues that I'm investigating.

**Comment by pieandcakes** [comment]
@jchadwick @isidorn This seems to be an issue still even with the latest insiders build.

**Comment by isidorn** [comment]
Reopening and assigning to may to investigate then

**Comment by pieandcakes** [comment]
I'm using VsCode Insiders 1.12.0 on OSX with the C++ extension. If you need a demo app to show this, I can share one with you.

**Comment by isidorn** [comment]
That would be great, could you share a demo app?

**Comment by pieandcakes** [comment]
@isidorn I sent it to you via email.

**Comment by isidorn** [comment]
@pieandcakes great! I'll check it out in the next couple of days

**Comment by isidorn** [comment]
@pieandcakes thanks for providing a reproducable repository. Providing such nice steps increases the probably that the bug will be fixed fast.

After investigation it seems like the initial value for the `allThreadsContinued` was set to `false`, and it should be `true`. I have fixed this and your demo app seems to work fine for me now.
You can try this out in tomorrow's insiders and let me know how it goes. 
Thanks

**Comment by chrmarti** [comment]
Unable to verify or reproduce. Tried and failed to launch the accepted example from https://stackoverflow.com/questions/266168/simple-example-of-threading-in-c . My launch config is:
```json
{
    "name": "(gdb) Launch",
    "type": "cppdbg",
    "request": "launch",
    "program": "${workspaceRoot}/a.out",
    "args": [],
    "stopAtEntry": false,
    "cwd": "${workspaceRoot}",
    "environment": [],
    "externalConsole": false,
    "MIMode": "gdb",
    "setupCommands": [
        {
            "description": "Enable pretty-printing for gdb",
            "text": "-enable-pretty-printing",
            "ignoreFailures": true
        }
    ]
}
```

I used `g++ -std=c++0x -pthread multithreading.cpp` on Ubuntu to compile. Lauching seems to fail early with: Unable to start debugging. Launch options string provided by the project system is invalid. The type initializer for 'System.Xml.Serialization.XmlSerializationReader' threw an exception. @pieandcakes

**Comment by pieandcakes** [comment]
@chrmarti Which OS are you trying it on and what version of VS? We're getting reports of this with our last update and i'm trying to track it down.

**Comment by chrmarti** [comment]
@pieandcakes This is Ubuntu 14.04, todays VSCode 1.13-insiders and ms-vscode.cpptools 0.11.3.

**Comment by chrmarti** [comment]
Leaving this for someone else to verify. The extension issue is tracked here: Microsoft/vscode-cpptools#792

**Comment by chrmarti** [comment]
Not sure if this changed with the fix, but I don't see the threads pause on the same breakpoint at the same time. They pause one after the other (the second only after the first continued), which doesn't seem to make sense, but maybe I'm using a bad example. I'm setting the breakpoint on the `cout` line:
```cpp
// g++ -std=c++0x -pthread -g multithreading.cpp
#include <string>
#include <iostream>
#include <chrono>
#include <thread>

using namespace std;

void runTask(string msg)
{
    cout << "Task says: " << msg << "\n";
}

int main()
{
    thread t1(runTask, "Hello 1");
    thread t2(runTask, "Hello 2");

    t1.join();
    t2.join();

    cout << "all done\n";
    std::this_thread::sleep_for(std::chrono::seconds(5));
}
```

Launch config:
```json
{
    "name": "(gdb) Launch",
    "type": "cppdbg",
    "request": "launch",
    "program": "${workspaceRoot}/a.out",
    "args": [],
    "stopAtEntry": false,
    "cwd": "${workspaceRoot}",
    "environment": [],
    "externalConsole": false,
    "MIMode": "gdb",
    "setupCommands": [
        {
            "description": "Enable pretty-printing for gdb",
            "text": "-enable-pretty-printing",
            "ignoreFailures": true
        }
    ]
},
```

One issue I see is that all three threads are decoracted with `PAUSED ON BREAKPOINT` when only one actually is (the second will pause after the first continued and the main thread doesn't pause at all):
![image](https://cloud.githubusercontent.com/assets/9205389/26790661/5c94ee9c-49c9-11e7-801a-bfa31178428d.png)

Reopening for clarifying what the expected behavior is and if any of this is a regression of this fix.

**Comment by isidorn** [comment]
This seems like an unrelated issue, moving to June and @pieandcakes can also clarify on what should be the expected behvaior here and if the original issue is fixed.

Due to the nature of the fix this can not be the regression of this fix.

**Comment by pieandcakes** [comment]
@chrmarti We always pause on all threads when we hit breakpoints. that's the nature of the C++ debugger. Are you saying that you are showing threads not paused when you hit breakpoints? 

@isidorn The original issue looks fixed with the latest insiders build .

**Comment by chrmarti** [comment]
@pieandcakes Please see my last https://github.com/Microsoft/vscode/issues/18195#issuecomment-306219697 .

**Comment by pieandcakes** [comment]
@chrmarti I saw your comment. I'm not understanding when you say:
> but I don't see the threads pause on the same breakpoint at the same time. They pause one after the other (the second only after the first continued), which doesn't seem to make sense, but maybe I'm using a bad example. 

Are you expecting them to hit the same breakpoint at exactly the same time on both threads? if so, then that isn't a good test because runtime is not deterministic. Testing with my multithreaded apps I haven't seen any strange behavior from stepping or pausing.

**Comment by chrmarti** [comment]
@pieandcakes The screenshot shows three threads with `PAUSED ON BREAKPOINT`, but only one of them shows a callstack. One issue with this is that only two of the threads are expected to hit the breakpoint, the main thread only joins these, but should never pause. The second issue is that while a second thread is expected to hit the breakpoint and pause, that thread (if it is indeed paused) does not show a callstack in the screenshot.

When I click 'Continue' that one thread with the callstack continues and immediately one of the other threads shows a callstack, again with all three threads showing `PAUSED ON BREAKPOINT`. One problem with this is that clearly not all threads are paused, despite the `PAUSED` indicator. A second problem is that this seems to go against what this issue was trying to address as I understand it: Both threads, one after the other (not simultaneously), should hit the breakpoint and pause and when I hit continue both should resume.

**Comment by pieandcakes** [comment]
@chrmarti 
> When I click 'Continue' that one thread with the callstack continues and immediately one of the other threads shows a callstack, again with all three threads showing  PAUSED ON BREAKPOINT . One problem with this is that clearly not all threads are paused, despite the  PAUSED  indicator. A second problem is that this seems to go against what this issue was trying to address as I understand it: Both threads, one after the other (not simultaneously), should hit the breakpoint and pause and when I hit continue both should resume.

All threads pause when a breakpoint is hit. As a breakpoint is hit, all threads stop so you can evaluate anything you want on any thread when a breakpoint is hit. 

if you enable `"logging": { "trace":true, "traceResponse":true }` you can see what VSCode sends and what we reply. 

For breakpoint stopped event, we send:
` stopped: {"threadId":7,"reason":"breakpoint","source":{"name":"threading.cpp","path":"Z:\\Testing\\kitchensink-4\\src\\threading.cpp","sourceReference":0},"line":12,"column":1,"text":null,"allThreadsStopped":true,"type":"stopped"}`

But then the stacktrace request we get always shows `threadId` of the one that was already expanded. I don't see more requests to request stackframes from us for the other threads that stopped. This seems like a bug in VSCode?

**Comment by isidorn** [comment]
@chrmarti @pieandcakes if there is an issue in VSCode can you please create a new issue with concise steps? Thank you.
I am closing this one since the original issue is verified, thanks for understanding.

**Comment by pieandcakes** [comment]
I opened #28198

**Comment by isidorn** [comment]
@chrmarti good job finding this issue, this let us uncover a candidate in #28198

**Comment by roblourens** [comment]
It seems like there is nothing to verify in this issue right now? Correct me if I'm wrong

**Comment by isidorn** [comment]
@roblourens yeah, this is covered by other issues / test plan items
